### PR TITLE
dcp-68 split metadata by technology type

### LINKED ISSUE #68: Bug/Enhancement: does not support more than 1 10X version in the same dataset
SCEA accept projects with different 10X versions (although they do not accept multiple scRNA-seq technology types). The hca2scea tool does not currently handle 2 10X versions in the same dataset. We need to update the tool to allow for this.

### PR DESCRIPTION
**Pull request by ami-day**
Functionality to split the input spreadsheet into multiple different inputs based on the number of unique technoloy types.
The SCEA team only allow 1 technology type per dataset (pair of MAGE-TAB files).

### CODE REVIEW

File: hca-to-scea/helpers/fetch_fastq_path.py [hunk 1]
```diff
@@ -158,30 +158,45 @@ def retrieve_xml_from_sra(run_accessions):
...
         paths_sra = retrieve_xml_from_sra(run_accessions)
     return paths_sra
 
+def get_sra_path_from_ena(study_accession, run_accessions):
+    paths_sra = {}
+    try:
+        request_url = f'http://www.ebi.ac.uk/ena/portal/api/filereport?accession={study_accession}&result=read_run&fields=run_accession,sra_ftp'
+        sra_results = pd.read_csv(request_url, delimiter='\t')
+        if sra_results.shape[0] > 0:
+            for i in range(0, len(list(sra_results['run_accession']))):
```
Comment (amnonkhen) [inline hca-to-scea/helpers/fetch_fastq_path.py]: If you need the list item as well as its index you could use `enumerate`

File: hca-to-scea/helpers/multitab_excel_to_single_txt.py [hunk 2]
```diff
@@ -1,9 +1,35 @@
...
+    xlsx_dict_tmp = copy.deepcopy(xlsx_dict)
+
+    '''Delete unused tabs.'''
+    for tab in xlsx_dict.keys():
+        if "protocol" in tab:
+            if xlsx_dict[tab].empty:
+                del xlsx_dict_tmp[tab]
+
+    xlsx_dict_tmp2 = copy.deepcopy(xlsx_dict_tmp)
+
```
Comment (amnonkhen) [inline hca-to-scea/helpers/multitab_excel_to_single_txt.py]: This bit should be a function of its own. You could name it like the comment.
We need to keep functions in the same level of abstraction and details, and not to overburden the reader with too many implementation details.

File: hca-to-scea/helpers/multitab_excel_to_single_txt.py [hunk 3]
```diff
@@ -1,9 +1,35 @@
...
+    '''Delete unused protocol keys.'''
+    biomaterials_tabs = ["specimen_from_organism","cell_line","organoid","cell_suspension"]
+    for tab in biomaterials_tabs:
+        if tab in xlsx_dict_tmp.keys():
+            protocol_ids = [protocol_id for protocol_id in xlsx_dict_tmp[tab].columns if ".protocol_core.protocol_id" in protocol_id]
+            for protocol_id in protocol_ids:
+                if xlsx_dict_tmp[tab][protocol_id].isna().all():
+                    del xlsx_dict_tmp2[tab][protocol_id]
```
Comment (amnonkhen) [inline hca-to-scea/helpers/multitab_excel_to_single_txt.py:30]: This also should be a function.

File: hca-to-scea/helpers/multitab_excel_to_single_txt.py [hunk 4]
```diff
@@ -46,57 +72,34 @@ def multitab_excel_to_dict(work_dir, excel_file):
...
-
-def merge_dataframes(xlsx_dict: {}) -> pd.DataFrame():
-
-    """Merge the sequence file df with the cell suspension df via the linked
-     cell suspension ids."""
+def merge_sample_types(xlsx_dict: {},experimental_design) -> pd.DataFrame():
 
     '''process.insdc_experiment.insdc_experiment_accession' is present in multiple tabs. Save the cell_suspension experiment accessions with a unique name.'''
-    xlsx_dict['cell_suspension']["cell_suspension.insdc_experiment.insdc_experiment_accession"] = list(xlsx_dict['cell_suspension']['process.insdc_experiment.insdc_experiment_accession'])
+    if 'process.insdc_experiment.insdc_experiment_accession' in xlsx_dict['cell_suspension'].columns:
```
Comment (amnonkhen) [inline hca-to-scea/helpers/multitab_excel_to_single_txt.py]: `xlsx_dict['cell_suspension']` repeats several times. Consider storing it in a variable.

File: hca-to-scea/helpers/utils.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        elif cell_line and organoid:
+            experimental_design = "organoid"
+        else:
+            experimental_design = "standard"
+
+    return experimental_design
+
+def check_for_pooled_samples(xlsx_dict):
+
+    df = xlsx_dict["specimen_from_organism"]
```
Comment (amnonkhen) [inline hca-to-scea/helpers/utils.py]: The name `df` is overloaded in this function for several uses. This is bad practice. Names should convey the intent of the writer as well as the purpose of the thing that is named.

File: hca-to-scea/helpers/utils.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
+#
```
Comment (amnonkhen) [inline hca-to-scea/helpers/utils.py:1]: `utils` is a name that suggests this file contains a number of things that are not related to one another. It means it would have many reasons to change, as opposed to a single one.
I see here "pure" calculation related functions, as well as ones that seem to be related to our domain.
In addition, the file is very long, which also suggests there is too much going on in this file.

File: hca-to-scea/helpers/utils.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+                        xlsx_dict_tmp[biomaterial_tab] = xlsx_dict[biomaterial_tab][
+                            xlsx_dict[biomaterial_tab]["{}.biomaterial_core.biomaterial_id".format(biomaterial_tab)].isin(
+                                (specimen_ids))]
+                if "donor_organism.biomaterial_core.biomaterial_id" in xlsx_dict_tmp[biomaterial_tab].columns:
+                    donor_ids_specimen = xlsx_dict_tmp[biomaterial_tab][
+                        "donor_organism.biomaterial_core.biomaterial_id"].values
+                    xlsx_dict_tmp["donor_organism"] = xlsx_dict["donor_organism"][
+                        xlsx_dict["donor_organism"]["donor_organism.biomaterial_core.biomaterial_id"].isin(
+                            (donor_ids_specimen))]
+
```
Comment (amnonkhen) [inline hca-to-scea/helpers/utils.py]: Too many nesting levels and branching.
This is untestable and will become unreadable, even by the author, in no time.

File: hca-to-scea/helpers/utils.py [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+                        protocol_id_list.extend(list(xlsx_dict_tmp[biomaterial_tab][id_column]))
+            protocol_id_list = list(set(protocol_id_list))
+            protocol_id_list = [protocol_id for protocol_id in protocol_id_list if pd.notnull(protocol_id)]
+            protocol_id_split = [protocol_id.split("||") for protocol_id in protocol_id_list if "||" in protocol_id]
+            if protocol_id_split:
+                protocol_id_split = [item for sublist in protocol_id_split for item in sublist]
+                protocol_id_split.extend([protocol_id for protocol_id in protocol_id_list if "||" not in protocol_id])
+                protocol_id_list = protocol_id_split
+            keep_protocols = [protocol_id for protocol_id in list(xlsx_dict_tmp[protocol_tab][id_column]) if protocol_id in protocol_id_list]
+            xlsx_dict_tmp[protocol_tab] = xlsx_dict_tmp[protocol_tab][xlsx_dict_tmp[protocol_tab][id_column].isin((keep_protocols))]
```
Comment (amnonkhen) [inline hca-to-scea/helpers/utils.py]: also too deep and too many branches

File: hca-to-scea/helpers/utils.py [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+
+    return xlsx_dict_tmp
+
+def check_technology_eligibility(library_method,technology_dict):
+
+    if library_method in technology_dict.keys():
+        eligible = True
+    else:
+        eligible = False
```
Comment (amnonkhen) [inline hca-to-scea/helpers/utils.py]:
```suggestion
    eligible =  library_method in technology_dict.keys()
```

**Review by amnonkhen — CHANGES_REQUESTED** [review]
This PR is way too big. It seem to include several changes, some functional, some related to repository cleanup. It is better to avoid this to facilitate simpler reviews and testing.
In addition, this contains changes to files that I have done, which should not appear hear. How did you update this branch with the latest changes?

**Comment by ami-day** [comment]
> This PR is way too big. It seem to include several changes, some functional, some related to repository cleanup. It is better to avoid this to facilitate simpler reviews and testing. In addition, this contains changes to files that I have done, which should not appear hear. How did you update this branch with the latest changes?

I think it appeared probably because I merged other branches into this pull request.
The pull request is big because I wanted to be able to test the multiple enhancements I made together, which I couldn't do without merging them. The relevant pull requests had not yet been merged into master but I knew they were working fine as I tested them.

File: hca-to-scea/helpers/utils.py (hunk 7 above)
Comment (ami-day) [reply hca-to-scea/helpers/utils.py]: I thought this too :-/ but I am not currently sure of a better way to do what I need the function to do. For now, I would like to leave it so it can be merged into master. And come back to it later. This is because I want us to be able to use the latest version of the tool, with the new functionality.

File: hca-to-scea/helpers/utils.py (hunk 8 above)
Comment (ami-day) [reply hca-to-scea/helpers/utils.py]: same comment as above

**Review by ami-day — COMMENTED** [review]
Thank you @amnonkhen , I really value this detailed feedback. I have made the suggested changes. Except, about the "too much nesting" comments:

- I would really like to merge this into master asap so we can use the new functionality. The latest version is much more robust against errors as it supports more complex experimental designs.

- It would be great to talk through how you would approach. refactoring the functions you have identified as too nested/long. I also have some questions about the best approach to decide when and how to split up functions/modules into separate files and how to name those files. I will set up a meeting if ok with you.

Thanks!
